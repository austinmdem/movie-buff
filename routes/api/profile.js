const express = require('express');
const router = express.Router();
const auth = require('../../middleware/auth');
const checkObjectId = require('../../middleware/checkObjectId');
const { check, validationResult } = require('express-validator');

const Profile = require('../../models/Profile');
const User = require('../../models/User');

// @route  GET api/profile/me
// @desc   Get current users profile
// @access Private
router.get('/me', auth, async (req, res) => {
  try {
    const profile = await Profile.findOne({
      user: req.user.id,
    }).populate('user', ['name', 'avatar']);

    if (!profile) {
      return res.status(400).json({ msg: 'There is no profile for this user' });
    }

    res.json(profile);
  } catch (err) {
    console.error(err.message);
    res.status(500).send('Server Error');
  }
});

// @route  POST api/profile/me
// @desc   Create or update user profile
// @access Private
router.post(
  '/',
  [
    auth,
    [
      check('favoritegenre', 'Favorite genre required').not().isEmpty(),
      check('topthreemovies', 'Top 3 movies required').not().isEmpty(),
    ],
  ],
  async (req, res) => {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({ errors: errors.array() });
    }

    const { topthreemovies, favoritegenre } = req.body;

    const profileFields = {
      user: req.user.id,
      favoritegenre,
      topthreemovies: Array.isArray(topthreemovies)
        ? topthreemovies
        : topthreemovies
            .split(',')
            .map((topthreemovie) => ' â€¢ ' + topthreemovie.trim()),
    };

    // Build profile object
    /* const profileFields = {};
    profileFields.user = req.user.id;
    if (favoritegenre) profileFields.favoritegenre = favoritegenre;
    if (topthreemovies) {
      profileFields.topthreemovies = topthreemovies
        .split(',')
        .map((topthreemovie) => topthreemovie.trim());
    }*/

    try {
      let profile = await Profile.findOne({ user: req.user.id });

      if (profile) {
        // Update
        profile = await Profile.findOneAndUpdate(
          { user: req.user.id },
          { $set: profileFields },
          { new: true }
        );

        return res.json(profile);
      }

      // Create
      profile = new Profile(profileFields);

      await profile.save();
      res.json(profile);
    } catch (err) {
      console.error(err.message);
      res.status(500).send('Server Error');
    }
  }
);

// @route  GET api/profile
// @desc   Get all profiles
// @access Public
router.get('/', async (req, res) => {
  try {
    const profiles = await Profile.find().populate('user', ['name', 'avatar']);
    res.json(profiles);
  } catch (err) {
    console.error(err.message);
    res.status(500).send('Server Error');
  }
});

// @route    GET api/profile/user/:user_id
// @desc     Get profile by user ID
// @access   Public
router.get(
  '/user/:user_id',
  checkObjectId('user_id'),
  async ({ params: { user_id } }, res) => {
    try {
      const profile = await Profile.findOne({
        user: user_id,
      }).populate('user', ['name', 'avatar']);

      if (!profile) return res.status(400).json({ msg: 'Profile not found' });

      return res.json(profile);
    } catch (err) {
      console.error(err.message);
      return res.status(500).json({ msg: 'Server error' });
    }
  }
);

// @route  DELETE api/profile
// @desc   Delete profile, user & posts
// @access Public
router.delete('/', auth, async (req, res) => {
  try {
    // Remove users posts
    await Post.deleteMany({ user: req.user.id });
    // Remove profile
    await Profile.findOneAndRemove({ user: req.user.id });
    await User.findOneAndRemove({ _id: req.user.id });
    res.json({ msg: 'User deleted' });
  } catch (err) {
    console.error(err.message);
    res.status(500).send('Server Error');
  }
});

module.exports = router;
