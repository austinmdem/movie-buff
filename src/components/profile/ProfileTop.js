import React from 'react';
import PropTypes from 'prop-types';

const ProfileTop = ({
  profile: {
    topthreemovies,
    favoritegenre,
    user: { name, avatar },
  },
}) => {
  return (
    <div className="profile-top bg-primary p-2">
      <img className="round-img my-1" src={avatar} alt="" />
      <h1 className="large">{name}</h1>
      <p className="lead">
        <p>{topthreemovies} hello</p>
      </p>
      <p>{favoritegenre}</p>
    </div>
  );
};

ProfileTop.propTypes = {
  profile: PropTypes.object.isRequired,
};

export default ProfileTop;
