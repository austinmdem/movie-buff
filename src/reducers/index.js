import { combineReducers } from 'redux';
import alert from './alert';
import auth from './auth';
import profile from './profile';
import post from './post';
import rating from './rating';

export default combineReducers({
  alert,
  auth,
  profile,
  post,
});
