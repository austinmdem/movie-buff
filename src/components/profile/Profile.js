import React, { Fragment, useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from '@reach/router';
import ProfileTop from './ProfileTop';
import Spinner from '../elements/Spinner';
import { getProfileById } from '../../actions/profile';

import { StyledProfile } from '../styles/StyledProfile';

const Profile = ({
  getProfileById,
  profile: { profile, loading },
  auth,
  id,
}) => {
  //
  useEffect(() => {
    getProfileById(id);
  }, [getProfileById, id]);

  return (
    <Fragment>
      {profile === null || loading ? (
        <Spinner />
      ) : (
        <StyledProfile>
          <Fragment>
            <div className="profile-nav">
              <Link to="/profiles" className="btn">
                Back to Profiles
              </Link>
              {auth.isAuthenticated &&
                auth.loading === false &&
                auth.user._id === profile.user._id && (
                  <Link to="/profile-form" className="btn">
                    Edit Profile
                  </Link>
                )}
              <ProfileTop profile={profile} />
            </div>
          </Fragment>
        </StyledProfile>
      )}
    </Fragment>
  );
};

Profile.propTypes = {
  getProfileById: PropTypes.func.isRequired,
  profile: PropTypes.object.isRequired,
  auth: PropTypes.object.isRequired,
};

const mapStateToProps = (state) => ({
  profile: state.profile,
  auth: state.auth,
});

export default connect(mapStateToProps, { getProfileById })(Profile);
