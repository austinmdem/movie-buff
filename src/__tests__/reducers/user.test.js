import * as types from '../../actions/types';
import * as reducer from '../../reducers/profile';

describe('Profile Reducer', () => {
  it('Should return the initial state', () => {
    expect(reducer.default(undefined, {})).toEqual({
      profile: null,
      profiles: [],
      loading: true,
      error: {},
    });
  });

  it('Should handle GET_PROFILE', () => {
    const action = {
      type: types.GET_PROFILE,
      payload: types.profile,
      loading: types.loading,
    };

    expect(reducer.default(undefined, action)).toMatchSnapshot();
  });

  it('Should handle GET_PROFILES', () => {
    const action = {
      type: types.GET_PROFILES,
      payload: types.profiles,
      loading: types.loading,
    };

    expect(reducer.default(undefined, action)).toMatchSnapshot();
  });

  it('Should handle PROFILE_ERROR', () => {
    const action = {
      type: types.PROFILE_ERROR,
      payload: types.error,
      loading: types.loading,
      profile: types.profile,
    };

    expect(reducer.default(undefined, action)).toMatchSnapshot();
  });

  it('Should handle CLEAR_PROFILE', () => {
    const action = {
      type: types.CLEAR_PROFILE,
      profile: types.profile,
    };

    expect(reducer.default(undefined, action)).toMatchSnapshot();
  });
});
