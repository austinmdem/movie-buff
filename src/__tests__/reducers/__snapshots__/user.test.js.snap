// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Reducer Should handle CLEAR_PROFILE 1`] = `
Object {
  "error": Object {},
  "loading": true,
  "profile": null,
  "profiles": Array [],
}
`;

exports[`Profile Reducer Should handle GET_PROFILE 1`] = `
Object {
  "error": Object {},
  "loading": false,
  "profile": undefined,
  "profiles": Array [],
}
`;

exports[`Profile Reducer Should handle GET_PROFILES 1`] = `
Object {
  "error": Object {},
  "loading": false,
  "profile": null,
  "profiles": undefined,
}
`;

exports[`Profile Reducer Should handle PROFILE_ERROR 1`] = `
Object {
  "error": undefined,
  "loading": false,
  "profile": null,
  "profiles": Array [],
}
`;
